src/GraphQL/Instrumentation/FieldResolverBuilderExtensions.cs

```diff
@@ -8,42 +8,72 @@ namespace GraphQL.Instrumentation
 {
     public static class FieldResolverBuilderExtensions
     {
-        private const string InvokeMethodName = "Resolve";
+        private const string INVOKE_METHOD_NAME = "Resolve";
 
-        public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) where T : new() => Use(builder, typeof(T));
+        public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder, IFieldMiddleware middleware)
+            => builder.Use(next => context => middleware.Resolve(context, next));
+
+        public static IFieldMiddlewareBuilder Use<T>(this IFieldMiddlewareBuilder builder) => Use(builder, typeof(T));
 
         public static IFieldMiddlewareBuilder Use(this IFieldMiddlewareBuilder builder, System.Type middleware)
         {
-            var methods = middleware.GetMethods(BindingFlags.Instance | BindingFlags.Public);
-            var invokeMethods = methods.Where(m => string.Equals(m.Name, InvokeMethodName, StringComparison.Ordinal)).ToArray();
-            if (invokeMethods.Length > 1)
+            // if IFieldMiddleware interface is supported, then just call its Resolve method
+            if (typeof(IFieldMiddleware).IsAssignableFrom(middleware))
             {
-                throw new InvalidOperationException($"There should be only a single method named {InvokeMethodName}. Middleware actually has {invokeMethods.Length} methods.");
+                return builder.Use(next =>
+                {
+                    return context =>
+                    {
+                        return context.Schema is Schema schema
```

Comment:
I know. This is the only place in PR that has been haunting me for some time. I understand that it would be possible to add the `ISchema.Services ` property, but I have always been against it. This motivates the use of a ServiceLocator pattern in client code. In addition, I do not think that such custom implementations exist at all. It’s much easier to inherit from `Schema`, not `ISchema`. Therefore, I use an explicit cast here.

Reply:
You're replacing one "anti-pattern" with another anti-pattern.  And it's not like exposing the Service property is going to help anyone -- it only provides access to the root service provider, and not the executing context's service provider.  Any singletons someone wants to use can be placed in the constructor of the graphtype.  Any scoped services still need to be accessed via an accessor.

Take a look at the Schema class.  Services doesn't even belong as a public member at all, unless to be exposed through ISchema, based on the 'anti-pattern' philosophy.  Just needs one change: `_lookup = new Lazy<GraphTypesLookup>(() => CreateTypesLookup(services));` and Schema.Services could be removed from the entire project.

I really don't think this PR should be pulled with this cast still in place.  I do think Services should be added to ISchema.  But if it isn't, then another solution should be found.

Reply:
> You're replacing one "anti-pattern" with another anti-pattern.

Where?

Reply:
`context.Schema is Schema schema`

Reply:
> Take a look at the Schema class. Services doesn't even belong as a public member at all, unless to be exposed through ISchema, based on the 'anti-pattern' philosophy. Just needs one change: _lookup = new Lazy<GraphTypesLookup>(() => CreateTypesLookup(services)); and Schema.Services could be removed from the entire project.

Good observation. It may be worth doing in order to be consistent until we find the best option.

> I really don't think this PR should be pulled with this cast still in place. I do think Services should be added to ISchema. But if it isn't, then another solution should be found.

let's consider an option _another solution should be found_.

Anyway i see it like this:

1. `Services` exists as an entity and the schema works with it to integrate with outer world.
2. So... in fact schema uses ServiceLocator pattern since schema is an infrastructure code not user code. In the end, some code MUST call `IServiceProvider.GetService()` and "get hands dirty".
3. User code should by all means avoid explicit work with `IServiceProvider.GetService()`. And we should not motivate user to call `IServiceProvider.GetService()` exposing `Schema.Services` or something else.

As a result: library code one way or another can work with `IServiceProvider`, but the user code is not.

Yes, and keep in mind that despite the fact that I do my best to protect the user code from the appearance of `Schema.Services`, sometimes I still think about giving it up and let it be.

Reply:
But I would start with the question "why do we need `ISchema` interface at all"? Is this really necessary?

Reply:
Good question.  I do not have the answer.

Reply:
I too. @joemcbride we need your help.

Reply:
How do you like this option?
```c#
public class Schema : ISchema, IServiceProvider
{
  object IServiceProvider.GetService(Type type) => _services.GetService(type);
}
```
\+Also remove `public IServiceProvider Services` from Schema and do this change:
```c#
  return context.Schema is IServiceProvider provider
 ? CheckNotNull((IFieldMiddleware)provider.GetService(middleware), middleware).Resolve(context, next)
                            : throw new NotSupportedException("Schema should implement IServiceProvoder interface");
```

So at least there may be custom `ISchema` implementations which could implement `IServiceProvider`.

Reply:
Well, it's a better option than before.  It is still poor code design; there's still a cast with a possibility of failure, when interfaces (which already exist) should protect against this.  In this example, ISchema should inherit from IServiceProvider.  (This of course then breaks the goal of hiding Services from the user.)  I still wouldn't do it.

Reply:
> It is still poor code design

In fact, I do not see the insoluble problems of this design. The only drawback of the inability to use custom (although I doubt that there are any) schemes we decided. It is enough for the client to implement the interface, a message about which he will see. And we (at least me) still not expose `Schema.Services`.

Reply:
> ISchema should inherit from IServiceProvider

I do not think so. Some interfaces are infrastructural in nature and are not intended for access by clients. At least clients should make some efforts and obtain access explicitly, that is, understand what they are doing.

Reply:
@Shane32 I followed your suggestion and removed `Schema.Services` property. In addition, I think it was dangerous to allow changing it after creating the schema.

Reply:
I believe the actual removal of the Schema.Services property should be in a separate PR.  While it does not affect any of my code, others may have differing points of view, and it would be good to allow time for comments before the removal.  It should not affect your implementation of this PR.

I would not object to the change, since if anyone wanted access to the root IServiceProvider, they could simply add a reference to it in the constructor of the graphtype.  Further, they could expose the property in their `MySchema` class if desired.  However, I certainly have used `((Schema)context.Schema).Services` in the past, and others may have similar code in their codebase.

> I think it was dangerous to allow changing it after creating the schema.

I see nothing wrong with making it a readonly property.

Reply:
((Schema)context.Schema).Services -> (IServiceProvider)context.Schema

I also thought about readonly property, but the fix is so simple that I decided to remove it right away.

Reply:
Makes sense

Reply:
> others may have differing points of view, and it would be good to allow time for comments before the removal

I'm in no hurry.